### PR TITLE
MAILBOX-320 Easy to instanciate Managers for tests <3

### PR DESCRIPTION
**Pull request by chibenwa**
This greatly reduces the uses of StoreMailboxManager related constructor and allows in tests to deal with them in a flexible way.

### CODE REVIEW

File: mailbox/tool/pom.xml [hunk 1]
```diff
@@ -48,12 +48,6 @@
         </dependency>
```
Comment (rouazana) [inline mailbox/tool/pom.xml]: s/depandancy/dependency/ in commit title

File: mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java [hunk 2]
```diff
@@ -1,164 +1,164 @@
...
+                for (int k=0; k < SUB_MAILBOXES_COUNT; k++) {
+                    
+                    String subFolderName = folderName + ".SUB_FOLDER_" + k;
+                    mailboxPath = MailboxPath.forUser(user, subFolderName);
+                    createMailbox(mailboxSession, mailboxPath);
+                    
+                    for (int l=0; l < SUB_SUB_MAILBOXES_COUNT; l++) {
+
+                        String subSubfolderName = subFolderName + ".SUBSUB_FOLDER_" + l;
+                        mailboxPath = MailboxPath.forUser(user, subSubfolderName);
```
Comment (rouazana) [inline mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java]: no need for variable reutilisation here

File: mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java [hunk 3]
```diff
@@ -1,164 +1,164 @@
...
+            
+        }
+        
+    }
+    
+    /**
+     * 
+     * @param mailboxPath
+     * @throws MailboxException
+     * @throws UnsupportedEncodingException
```
Comment (rouazana) [inline mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java]: useless and wrong javadoc

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/MailboxManagerOptions.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+package org.apache.james.mailbox.store;
+
+import java.util.Optional;
+
+import org.apache.james.mailbox.acl.GroupMembershipResolver;
+import org.apache.james.mailbox.acl.SimpleGroupMembershipResolver;
+import org.apache.james.mailbox.model.MailboxConstants;
+import org.apache.james.mailbox.store.mail.model.impl.MessageParser;
+
+public class MailboxManagerOptions {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/MailboxManagerOptions.java]: Options?

Comment (rouazana) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/MailboxManagerOptions.java]: BadNamingException? :) The naming service is unreachable, and I have no better idea for the moment.

File: mailbox/api/src/test/java/org/apache/james/mailbox/manager/ManagerTestResources.java [hunk 5]
```diff
@@ -71,7 +71,7 @@ public ManagerTestResources(IntegrationResources<T> integrationResources) throws
         quotaRootResolver = integrationResources.createQuotaRootResolver(mailboxManager);
         quotaManager = integrationResources.createQuotaManager(maxQuotaManager, mailboxManager);
         integrationResources.init();
-        session = mailboxManager.login(USER, USER_PASS);
+        session = mailboxManager.createSystemSession(USER);
```
Comment (rouazana) [inline mailbox/api/src/test/java/org/apache/james/mailbox/manager/ManagerTestResources.java]: I don't understand why it's better: if you login, you are reproducing a real behaviour, while creating a system session you are not. No?

File: mpt/impl/imap-mailbox/maildir/src/test/java/org/apache/james/mpt/imapmailbox/maildir/host/MaildirHostSystem.java [hunk 6]
```diff
@@ -27,67 +27,49 @@
...
     private static final String MAILDIR_HOME = "target/Maildir";
     private static final ImapFeatures SUPPORTED_FEATURES = ImapFeatures.of();
     
+    private final TemporaryFolder temporaryFolder;
     private StoreMailboxManager mailboxManager;
 
-    public static JamesImapHostSystem build() throws Exception {
-        return new MaildirHostSystem();
+    public MaildirHostSystem(TemporaryFolder temporaryFolder) {
+        this.temporaryFolder = temporaryFolder;
```
Comment (rouazana) [inline mpt/impl/imap-mailbox/maildir/src/test/java/org/apache/james/mpt/imapmailbox/maildir/host/MaildirHostSystem.java]: unrelated change?

File: mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/MaildirMailboxManagerProvider.java [hunk 7]
```diff
@@ -21,41 +21,36 @@
...
 import org.apache.james.mailbox.store.event.DefaultDelegatingMailboxListener;
 import org.apache.james.mailbox.store.event.MailboxEventDispatcher;
 import org.apache.james.mailbox.store.mail.model.DefaultMessageId;
-import org.apache.james.mailbox.store.mail.model.impl.MessageParser;
 import org.junit.rules.TemporaryFolder;
 
 public class MaildirMailboxManagerProvider {
-    public static StoreMailboxManager createMailboxManager(String configuration, TemporaryFolder temporaryFolder) throws MailboxException, IOException {
+    public static StoreMailboxManager createMailboxManager(String configuration, TemporaryFolder temporaryFolder,
+                                                           MailboxManagerOptions options) throws MailboxException, IOException {
```
Comment (rouazana) [inline mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/MaildirMailboxManagerProvider.java]: what's the point of this option stuff if it's still adding some parameters to methods, instead of removing?

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 8]
```diff
@@ -61,6 +61,9 @@
  */
 public abstract class MailboxManagerTest {
 
+    protected static final int LIMIT_ANNOTATIONS = 3;
+    protected static final int LIMIT_ANNOTATION_SIZE = 30;
```
Comment (rouazana) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: never used?

Comment (chibenwa) [reply mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: Will be used by memory and cassandra who for now duplicates them

File: mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java]: It's just a class rename. Should I refactor it?

File: mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java (hunk 3 above)
Comment (chibenwa) [reply mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java]: Idem.

File: mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/MaildirMailboxManagerProvider.java (hunk 7 above)
Comment (chibenwa) [reply mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/MaildirMailboxManagerProvider.java]: The goal is to say: now I can instanciate  whatever maildir manager with whatever componant configued, without :
 - introducing several provider methods
 - changing 1000  constructors

Note that this change allow me to use the provider in mpt tests and avoid a lot of burden.

The idea of this commit is to remove coupling. Coupling between every test that uses a manager and the constructor itself. Now when I change the constructor of the manager, I need to change a way more limited number of class (test class not using the given added parameter will not need to change, at all!)

File: mpt/impl/imap-mailbox/maildir/src/test/java/org/apache/james/mpt/imapmailbox/maildir/host/MaildirHostSystem.java (hunk 6 above)
Comment (chibenwa) [reply mpt/impl/imap-mailbox/maildir/src/test/java/org/apache/james/mpt/imapmailbox/maildir/host/MaildirHostSystem.java]: No not at all!!!

I want to use that provider in MPT!!!! That's the aim of that commit.

The clue you miss is that we could not rely on junit rule with the previous version of imap mot test prior @mbeachler refactoring. So we did a hack to still be able to run them.

Now we can ;-) and doing it is simple. It allows using the provider.

File: mailbox/api/src/test/java/org/apache/james/mailbox/manager/ManagerTestResources.java (hunk 5 above)
Comment (chibenwa) [reply mailbox/api/src/test/java/org/apache/james/mailbox/manager/ManagerTestResources.java]: Well, no need to customize authentication logic and add a fake user when it is only used in test setUp?

File: mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java [hunk 9]
```diff
@@ -1,164 +1,164 @@
...
+
+import org.apache.james.mailbox.MailboxManager;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.MessageManager;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.mock.MockMail;
+import org.apache.james.mailbox.model.MailboxPath;
+
+/**
+ * A mock mailbox manager.
```
Comment (mbaechler) [inline mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java]: or not

File: mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java [hunk 10]
```diff
@@ -1,164 +1,164 @@
...
+ *
+ */
+public class MailboxManagerFeeder {
+    
+    /**
+     * The mock mailbox manager constructed based on a provided mailboxmanager.
+     */
+    private final MailboxManager mockMailboxManager;
```
Comment (mbaechler) [inline mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java]: mock ?

File: mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java (hunk 2 above)
Comment (mbaechler) [reply mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java]: Sorry but renaming a "mock"  thing to a "feeder" thing can't come without any code change, a lot of assumption become wrong

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 11]
```diff
@@ -214,10 +214,12 @@ public void closingSessionShouldWork() throws MailboxException, UnsupportedEncod
 
     @Test
     public void listShouldReturnMailboxes() throws MailboxException, UnsupportedEncodingException {
+        new MailboxManagerFeeder(mailboxManager).feed();
```
Comment (mbaechler) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: it should be `MailboxManagerFeeder.feed(mailboxManager)` in my opinion

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/MailboxManagerOptions.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+                groupMembershipResolver.orElse(new SimpleGroupMembershipResolver()),
+                messageParser.orElse(new MessageParser()));
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static final MailboxManagerOptions NONE = builder().build();
```
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/MailboxManagerOptions.java]: it's default, not none

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/MailboxManagerOptions.java (hunk 4 above)
Comment (mbaechler) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/MailboxManagerOptions.java]: it could be a configuration but it mixes different concerns : there are implementation (IoC like) and configuration values.
IMO it's the sign it should be two different things

File: mailbox/hbase/src/test/java/org/apache/james/mailbox/hbase/HBaseMailboxManagerProvider.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+import org.apache.james.mailbox.store.StoreMailboxAnnotationManager;
+import org.apache.james.mailbox.store.StoreRightManager;
+import org.apache.james.mailbox.store.event.DefaultDelegatingMailboxListener;
+import org.apache.james.mailbox.store.event.MailboxEventDispatcher;
+import org.apache.james.mailbox.store.mail.model.DefaultMessageId;
+
+import com.google.common.base.Throwables;
+
+public class HBaseMailboxManagerProvider {
+    public MailboxManager provideMailboxManager(HBaseClusterSingleton cluster, MailboxManagerOptions options) {
```
Comment (mbaechler) [inline mailbox/hbase/src/test/java/org/apache/james/mailbox/hbase/HBaseMailboxManagerProvider.java]: Ok, I think I can come up with a better design : each backend could provide a MailboxManager.Builder (for test only) that would allow building a default MailboxManager but still allow customization for tests in need of it.
Looks fine to you ?

File: mailbox/api/src/test/java/org/apache/james/mailbox/manager/ManagerTestResources.java (hunk 5 above)
Comment (mbaechler) [reply mailbox/api/src/test/java/org/apache/james/mailbox/manager/ManagerTestResources.java]: if IRL people get sessions from login method, we should use login method in test.

File: mailbox/hbase/src/test/java/org/apache/james/mailbox/hbase/HBaseMailboxManagerProvider.java (hunk 13 above)
Comment (chibenwa) [reply mailbox/hbase/src/test/java/org/apache/james/mailbox/hbase/HBaseMailboxManagerProvider.java]: That's a lot of boilerplate code, repeating that builder everywhere.

And every time we do a code change to that constructor, we would need to modify each builder...

In my opinion we are witnessing the consequences of inheritance (as opposed to  composition). After all I concider we are not far from a single storeMailboxManager class. We just need to compose:
 - instanciation of MessageMapper
 - capabilities

I would then prefer invest in that direction. At least it fixes things at the root.

Comment (mbaechler) [reply mailbox/hbase/src/test/java/org/apache/james/mailbox/hbase/HBaseMailboxManagerProvider.java]: > That's a lot of boilerplate code, repeating that builder everywhere.

I can't understand why. Retrieving the default one is as short as what you propose.

> And every time we do a code change to that constructor, we would need to modify each builder...

Yes, 5 or 6 builder to change for a new parameter to MailboxManager. What's the problem ? Do you expect MailboxManager to have an ever growing number of dependencies ? I wish not.

Finally, about StoreMailboxManager, It's another issue, let's manage the current one here.

Comment (chibenwa) [reply mailbox/hbase/src/test/java/org/apache/james/mailbox/hbase/HBaseMailboxManagerProvider.java]: > Finally, about StoreMailboxManager, It's another issue, let's manage the current one here.

No but reorganizing concepts in it requires changing constructor. And right now, it's a hell. Cc @strokyl  @quynhn  .

**Comment by aduprat** [comment]
Build triggered. sha1 is merged.

**Comment by aduprat** [comment]
Build started sha1 is merged.

**Comment by aduprat** [comment]
Build finished. No test results found.

**Comment by aduprat** [comment]
Build triggered. sha1 is merged.

**Comment by aduprat** [comment]
Build started sha1 is merged.

**Comment by aduprat** [comment]
Build finished. No test results found.

**Comment by aduprat** [comment]
Build triggered. sha1 is merged.

**Comment by aduprat** [comment]
Build started sha1 is merged.

**Comment by aduprat** [comment]
Build finished. No test results found.

File: mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java (hunk 9 above)
Comment (strokyl) [reply mailbox/api/src/test/java/org/apache/james/mailbox/manager/MailboxManagerFeeder.java]: :+1:

**Comment by openpaas-james** [comment]
Build triggered. sha1 is original commit.

**Comment by openpaas-james** [comment]
Build started sha1 is original commit.

**Comment by openpaas-james** [comment]
Build finished. No test results found.

**Comment by chibenwa** [comment]
No longer relevant (inMemory utilities ease this job a lot)
